src/doc/doc-printer.js

```diff
@@ -4,7 +4,7 @@ const { getStringWidth } = require("../common/util");
 const { convertEndOfLineToChars } = require("../common/end-of-line");
 const { concat, fill, cursor } = require("./doc-builders");
 
-/** @type {{[groupId: PropertyKey]: MODE}} */
+/** @type {{[groupId: string]: number}} */
```

Comment:
```suggestion
/** @type {Record<symbol, GroupMode>} */
```

Reply:
Turns out TS doesn't really support symbol indexers: https://github.com/microsoft/TypeScript/issues/1863
It'll be an error if we set `noImpliciyAny` to `true`. But that's how this variable is used. Its keys are symbols, values 1 or 2. Defining its as `{[groupId: string]: number}` is a mistake.

Reply:
Thanks @thorn0, I committed the suggestion. I must admit that this is something that I don't understand too well. But I am really happy that we can get this fixed.

Reply:
What exactly do you not understand?

Reply:
How you used Record, symbol, typeof MODE_BREAK, and typeof MODE_FLAT.

Looks nice and clean to me, though.

I guess I gotta to study TypeScript a bit more.

(I wouldn't mind a quick link, if you have one)

Reply:
* `Record` is a [predefined utility type](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkt).
* [Symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) aren't a TS thing, it's a primitive value type in JS. Their purpose is to represent unique keys that never clash. Unfortunately, TS's type system doesn't fully support them (see the link to the issue in my other message).
* The [`typeof` operator](https://basarat.gitbooks.io/typescript/docs/types/moving-types.html#capturing-the-type-of-a-variable) applied to a constant returns a literal type.